File: 2020/day3.rb
GRID = File.read('day3.txt').split("\n").freeze
ROW_LENGTH = 31

def traverse(y, x)
  trees = 0
  position = [0, 0]

  loop do
    break if position.first >= GRID.length

    GRID[position.first][position.last] == '#' && trees += 1
    position[0] += y
    position[1] = (position[1] + x) % ROW_LENGTH
  end

  trees
end

p traverse(1, 3)
p [[1, 1], [1, 3], [1, 5], [1, 7], [2, 1]].reduce(1) { |acc, arr| traverse(*arr) * acc }

File: 2020/day3.exs
defmodule Day3 do
  @grid File.stream!('day3.txt') |> Enum.map(&String.trim/1)
  @grid_length @grid |> Enum.count()
  @grid_width @grid |> hd |> String.length()

  def part1 do
    traverse({0, 0}, 0, 1, 3)
  end

  def part2 do
    [{1, 1}, {1, 3}, {1, 5}, {1, 7}, {2, 1}]
    |> Enum.map(fn {y, x} -> Task.async(fn -> traverse({0, 0}, 0, y, x) end) end)
    |> Task.await_many
    |> Enum.reduce(&(&1 * &2))
  end

  def traverse({pos_y, _pos_x}, trees, _y, _x) when pos_y >= @grid_length do
    trees
  end

  def traverse({pos_y, pos_x}, trees, y, x) do
    case @grid |> Enum.at(pos_y) |> String.at(pos_x) do
      "." ->
        traverse({pos_y + y, rem(pos_x + x, @grid_width)}, trees, y, x)

      "#" ->
        traverse({pos_y + y, rem(pos_x + x, @grid_width)}, trees + 1, y, x)
    end
  end
end

Day3.part1() |> IO.puts()
Day3.part2() |> IO.puts()
